### ISSUE TITLE
Default adam epsilon to 1e-7 when on fp16

### ISSUE THREAD

**Issue by williamFalcon** [pytorch/pytorch#26218]
## 🚀 Feature
If you use 1e-8 as default and you use 16 bit, it will round to zero, and thus you get no stability gains.

**Comment by mruberry** [comment]
@mcarilli @csarofeen, who work on mixed precision.

**Comment by mcarilli** [comment]
I'm a little wary of this.  When @puririshi98 was running some experiments training networks with fp16 parameter values + stochastic rounding, at one point we ran into numerical instabilities and one of the things we tried was increasing the value of epsilon.  The numerical instabilities turned out to be caused by something completely unrelated to epsilon, but I recall that increasing epsilon in some cases impaired the converged accuracy, even for pure fp32 runs.  I can dig up the exact networks and results if needed.

In general the optimizer shouldn't update fp16 parameters directly, because small updates later in training won't be captured.  This is also true for bfloat16 (even more so in fact, because bfloat has fewer mantissa bits), and is totally independent from the value of epsilon.  Given that the optimizer should act on fp32 parameters, and that (in our anecdotal experience) increasing epsilon can impair accuracy, to be safe I lean towards keeping the value of epsilon where it is, even if 1e-8 flushes to zero in fp16.

**Comment by williamFalcon** [comment]
i could see that. Using 1e-7 has made a huge different for my training on fp16 - especially with NCE loss. The 1e-8 WILL become 0 on fp16 (you can run a quick test by converting that value into .half() to verify). 

i’m worried that a lot of people using fp16 abandon projects because of stability issues likely caused by a default no one messes with. 

so, if 1e-8 = 0 when using half, it’s basically not serving any purpose as a default

**Comment by soumith** [comment]
this seems like a reasonable default for fp16, especially if someone does the updates in fp16 (rather naively).

@williamFalcon will accept a PR to fix this.

**Comment by f2013519** [comment]
@williamFalcon  Why 1e-7 in particular? For some loss methods, even this is not sufficient: https://github.com/keras-team/keras/issues/11860. Maybe 1e-4 instead?

Also will this be the default for both "pure" fp16 training and the training which uses mixed precision techniques (like master copy in fp32)? 

It might impact accuracy for the latter case and since the general recommendation is to do training with mixed precision (as opposed to pure half) due to reasons mentioned by @mcarilli above,  this might be silently running into accuracy loss as it might not be apparent to users that eps. value could be contributing to the loss.  This might be harder to debug than seeing inf/nan loss during training.

**Comment by mcarilli** [comment]
For parameters that are fp16, i guess it makes sense to increase epsilon to a value that's > 0 in fp16. But imo it doesn't make sense to train with pure fp16 parameters at all, at least not without stochastic rounding or some other fancy tricks that Pytorch optimizers don't currently implement.

My main recommendation is that when acting on fp32 params, we don't increase epsilon above its current value of 1e-8.

**Comment by f2013519** [comment]
If the change is for fp16 params only, it should be fine. The only question is what should be the exact value , 1e-7, 1e-4, something else?

**Comment by williamFalcon** [comment]
i chose 1e-7 because it is the smallest value which won’t become zero. I’m hesitant to use a larger value because it might bias gradients too much. My intuition here is to use the minimum value that won’t become zero.

**Comment by AAnoosheh** [comment]
Is this still an issue?  Because Adam's eps is still 1e-8.

Or does Lightning not actually perform backprop/updates in 16-bit?

**Comment by csarofeen** [comment]
CC: @xwang233 @ptrblck

**Comment by ptrblck** [comment]
@williamFalcon If I understand the issue correctly, you are using pure FP16 training by calling `model.half()` in PyTorch Lightning and suggest to increase the `eps` values for `Adam` for these FP16 parameters or are you seeing the same issue using `amp`?
